### PR TITLE
Laboratory work #2, Aleksandra Doroshina - 23FPL1

### PR DESCRIPTION
**Pull request by DoroshinaAleksandra**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,6 +21,16 @@ def tokenize(text: str) -> list[str] | None:
...
     """
+    if not isinstance(text, str):
+        return None
+    tokenized_text_str = ''
+    for element in text.lower():
+        if element.isalpha() or element == ' ':
+            tokenized_text_str += element
+        else:
+            tokenized_text_str += ' '
+    tokenized_text = tokenized_text_str.split()
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:32]: do not create the variable

File: lab_2_retrieval_w_bm25/settings.json [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 6
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/settings.json:2]: what about 8?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (DoroshinaAleksandra) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -47,6 +70,19 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+            or not documents):
+        return None
+    for text in documents:
+        for word in text:
+            if not isinstance(word, str):
+                return None
+    unique_words = []
+    for text in documents:
+        for word in text:
+            if word not in unique_words:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:83]: use set

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -47,6 +70,19 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+    if (not isinstance(documents, list) or not all(isinstance(text, list) for text in documents)
+            or not documents):
+        return None
+    for text in documents:
+        for word in text:
+            if not isinstance(word, str):
+                return None
+    unique_words = []
+    for text in documents:
+        for word in text:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:82]: convert text to set without loop

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (DoroshinaAleksandra) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (DoroshinaAleksandra) [reply lab_2_retrieval_w_bm25/main.py]: done

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -18,6 +21,15 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+    tokenized_text = ''
+    for element in text.lower():
+        if element.isalpha() or element == ' ':
+            tokenized_text += element
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:29]: `continue` and remove `else`

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -33,6 +45,17 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
     In case of corrupt input arguments, None is returned.
     """
+    if (not isinstance(tokens, list) or not all(isinstance(token, str) for token in tokens)
+            or not tokens):
+        return None
+    if (not isinstance(stopwords, list) or not all(isinstance(words, str) for words in stopwords)
+            or not stopwords):
+        return None
+    new_tokens = copy.deepcopy(tokens)
+    for token in tokens:
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:55]: Please find a new way without copy and remove

File: lab_2_retrieval_w_bm25/start.py [hunk 7]
```diff
@@ -26,8 +29,37 @@ def main() -> None:
...
+    if not isinstance(vocabulary, list):
+        return None
+    tf_idf = []
+    bm25 = []
+    idf = calculate_idf(vocabulary, clear_texts)
+    avgdl = sum(len(doc) for doc in documents) / len(documents)
+    for text in clear_texts:
+        tf = calculate_tf(vocabulary, text)
+        if not isinstance(tf, dict) or not isinstance(idf, dict):
+            return None
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/start.py:51]: remove `None`, just `return`

File: lab_2_retrieval_w_bm25/start.py [hunk 8]
```diff
@@ -26,8 +29,37 @@ def main() -> None:
...
+        bm25_text = calculate_bm25(vocabulary, text, idf, 1.5, 0.75, avgdl, len(text))
+        if not isinstance(tf_idf_text, dict) or not isinstance(bm25_text, dict):
+            return None
+        tf_idf.append(tf_idf_text)
+        bm25.append(bm25_text)
+    rank_tf_idf = rank_documents(tf_idf, 'Which fairy tale has Fairy Queen?', stopwords)
+    rank_bm25 = rank_documents(bm25, 'Which fairy tale has Fairy Queen?', stopwords)
+    result = rank_tf_idf, rank_bm25
     assert result, "Result is None"
+    return None
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/start.py:62]: remove

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**